Blog: The benefits, use cases, and types of feature flags explained (#5028)
PostHog/posthog.com: contents/blog/feature-flag-benefits-use-cases.md

---
date: 2023-01-13
title: Feature flag benefits, types and use cases, explained
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
author: ["ian-vanagas"]
featuredImage: ../images/blog/posthog-marketing/marketing-hog.jpg
featuredImageType: full
categories: Using PostHog
tags:
  - Feature flags
---

## What is a feature flag?

Feature flags (aka feature toggles) are a powerful tool to help improve your product. They enable engineering teams to conditionally run code, display components, and control access based on release conditions. These conditions check user data, group data, and percentages to decide whether to activate or not.

There are three broad categories of of feature flags:

- **Release (Beta, Canary):** Enable a feature for a percentage or category of users or groups, and monitor the impact. Roll back if there are issues. Roll out to everyone once ready.

- **Permission:** Grant or deny access to a feature to specific users or groups.

> - **Experiment (A/B):** Compare two or more versions of a feature or component for users or groups, and monitor the impact the variations have on a set of key metrics. Roll out the winning variation.

Reviewer: Moved this up here as it seemed to make more sense

Complex permission logic isn’t needed, engineering teams can just use a feature flag. This simplifies and distributes permission and access management. This means that a sales, support, or customer success team can then add permissions to their workflows and responsibilities. This takes work off the plate of developers. 

Feature flags also enable you to run alphas and betas of new features easily. Each feature doesn’t need a bespoke change to the code, other than the feature itself. The type of release or access pattern can change easily. For example, it can go from an alpha to a specific group release to a percentage rollout without needing to change any code.

#### Use cases

- Put premium features behind a permission-type feature flag. Once users sign up, get the sales or customer success team to add them to the feature flag distribution. If they churn, remove them from the feature flag release conditions.

- A user has been abusing your app. They have been sending too many requests or using the feature inappropriately. You can use a permission-type feature flag to prevent or throttle access to key parts of your app from them.

> ## Comparing feature flag tools

Reviewer: ```suggestion
## Choosing a feature flag tool
```

---

> date: 2023-01-09

Reviewer: ```suggestion
date: 2023-01-12
```

> title: The benefits, types, and use cases of feature flags explained

Reviewer: ```suggestion
title: Feature flag benefits, types and use cases, explained
```

Better this way around for SEO

Author: Include an oxford comma?

rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
author: ["ian-vanagas"]
featuredImage: ../images/blog/posthog-marketing/marketing-hog.jpg
featuredImageType: full

> categories: ["Guides", "Using PostHog"]

Reviewer: ```suggestion
categories: Using PostHog
tags:
  - Feature flags
```

---

> Feature flags (aka feature toggles) are a powerful tool to help improve your product. They enable engineering teams to conditionally run code, display components, and enable access based on release conditions. These conditions check user data, group data, and percentages to decide whether to activate or not.

Reviewer: ```suggestion
## What is a feature flag?

Feature flags (aka feature toggles) are a powerful tool to help improve your product. They enable engineering teams to conditionally run code, display components, and control access based on release conditions. These conditions check user data, group data, and percentages to decide whether to activate or not.
```

Change to avoid repetition. Is 'control' a suitable substitute here?

Also adding the heading at the top of the article for SEO. Might help us get a snippet.

In this post, we explore four benefits of feature flags along with use cases and types for each. The four biggest benefits and uses of feature flags are:
1. [shipping faster](#1-ship-faster-and-maintain-momentum)
2. [mitigating risk](#2-mitigate-risk)
3. [testing changes](#3-test-changes-in-production)
4. [managing access](#4-manage-access)

> 
> ## 1. Ship faster and maintain momentum
> 

Reviewer: ```suggestion

## Feature flag benefits and use cases

### 1. Ship faster and maintain momentum

```

> Feature flags enable developers to ship code faster. They do this by disconnecting deployment from release. Instead of releasing code for use by everyone once deployed, it lives behind a feature flag and releases separately. This frees developers from worrying as much about releases.

Reviewer: ```suggestion
Feature flags disconnect deployment from release, enabling developers to ship code faster. Instead of releasing code for use by everyone once deployed, it lives behind a feature flag and releases separately. This frees developers from worrying as much about releases.
```

Using feature flags improves the deployment and integration process. Feature flags help developers merge and integrate their changes more often. Because developers aren’t waiting to deploy, metrics like deployment frequency and time improve. You avoid stale branches and code because it isn’t sitting dormant. All of these benefit DevOps teams especially. 

Engineers can also use feature flags to hand off the release to other members of their team, such as product managers and marketers. This can be especially useful for features requiring slow rollouts, communication, and usage analysis. Instead of a developer babysitting the release, a non-technical team member can do the release, and any related work, themselves.

Feature flags also prevent the loss of momentum from issues with new features. If a rollback needs to happen, it can happen quickly and the team can maintain its momentum. Multiple team members aren’t needed to roll back a feature, it just requires turning off the flag.

> ### Use cases

Reviewer: ```suggestion
#### Use cases
```

> - You identify a major bug with a new feature. The developer who shipped the feature turns off the feature flag. They then complete a quick fix and ship it behind a bug-fix-type feature flag. They then continue to improve and clean up the bug fix and roll it out further once ready.
> - The team is planning the release of a big feature. It requires testing, a tiered rollout, customer communication, and marketing. The developer completes the feature, and puts it behind a release-type feature flag. They hand off the release to the product manager who owns the feature. The developer can get back to work while the PM coordinates with QA, customer success, and marketing on a rollout.

Reviewer: ```suggestion
- You identify a major bug with a new feature. The developer who shipped the feature turns off the feature flag. They then complete a quick fix and ship it behind a bug-fix-type feature flag. They then continue to improve and clean up the bug fix and roll it out further once ready.

- The team is planning the release of a big feature. It requires testing, a tiered rollout, customer communication, and marketing. The developer completes the feature, and puts it behind a release-type feature flag. They hand off the release to the product manager who owns the feature. The developer can get back to work while the PM coordinates with QA, customer success, and marketing on a rollout.
```

Longer bullet points need a line break between them to aid readability.

> ## 2. Mitigate risk

Reviewer: ```suggestion
### 2. Mitigate risk
```

By doing percentage, conditional, or tiered deployments, the rollout slows down. Feature flags can limit a rollout to specific infrastructure, regions, users, or groups instead of everyone, everywhere, all at once. Teams can focus on monitoring the rollout and its impact on specific areas and analytics. As a rollout progresses, risk decreases and the team can be more confident the feature won’t cause issues for users. 

Slower rollouts provide time for problems to arise before everyone accesses a feature. Using conditions (like specific users or groups) allows for discussion about changes and any issues they are seeing. This is also a way to get feedback about the feature (which can limit 
”business risk”). This ensures confidence in a feature by the time it rolls out to everyone.

If there are problems with a feature, feature flags can act as kill switches. There is no need to frantically make changes to remove the feature and then deploy those changes (which can lead to more problems). Just turn the feature flag off, preventing further issues, and you get time to analyze what went wrong.

> ### Use cases

Reviewer: ```suggestion
#### Use cases
```

> - Reports come in about a problem with a feature. Luckily, it is behind a release-type feature flag. This enables you to quickly roll back the feature, identify the issue, make a fix, and ship it. This is much faster than reverting the change and deploying an update.
> - You’re in a situation where every change to your app needs a review to release (such as iOS App Store or Chrome Web Store). Instead of having to wait for a review to make a change, using a feature flag enables faster rollbacks. Fixes to issues aren’t delayed by review, mitigating their impact when they arise.

Reviewer: ```suggestion
- Reports come in about a problem with a feature. Luckily, it is behind a release-type feature flag. This enables you to quickly roll back the feature, identify the issue, make a fix, and ship it. This is much faster than reverting the change and deploying an update.

- You’re in a situation where every change to your app needs a review to release (such as iOS App Store or Chrome Web Store). Instead of having to wait for a review to make a change, using a feature flag enables faster rollbacks. Fixes to issues aren’t delayed by review, mitigating their impact when they arise.
```

> 

Reviewer: Can you add a short boxout at the end of this section that defines percentage, conditional and tiered deployments?

> ## 3. Test changes (in production)

Reviewer: ```suggestion
### 3. Test changes (in production)
```

Feature flags enable better testing of changes by testing them closer to reality. When a change ships into production behind a feature flag and people who need access can get access. Instead of needing specific testing environments, you can test in production.

Testing in production means the usage, data, infrastructure, integration, and environment are closer to what users actually use. The gap between development and production closes. Other team members can test your code and features in production as well. All of this provides validation to engineers that their code works.

Testing changes in production also means you can run experiments. You can test variations against each other to see which provides better results. Product and marketing teams can be involved in the experimentation process. They can design experiments and analyze real results, rather than guessing what the outcomes are.

> PostHog has a built-in [experimentation suite](/product/experimentation-suite) that connects to all your product data and feature flags.

> ### Use cases

Reviewer: ```suggestion
#### Use cases
```

> - To test multiple variations of a pricing table with different information, you set up an experiment with multiple flag variants and track conversion on that page. At the end of the experiment, roll out the winning variant and roll back the losing ones.
> - Have a feature but you aren’t sure how it integrates in production. Ship it behind a test-type feature flag, and set distribution to yourself. If a teammate wants to test it too, you can add them to the release conditions as well.

Reviewer: ```suggestion
- To test multiple variations of a pricing table with different information, you set up an experiment with multiple flag variants and track conversion on that page. At the end of the experiment, roll out the winning variant and roll back the losing ones.

- Have a feature but you aren’t sure how it integrates in production. Ship it behind a test-type feature flag, and set distribution to yourself. If a teammate wants to test it too, you can add them to the release conditions as well.
```

> ## 4. Manage access

Reviewer: ```suggestion
### 4. Manage access
```

> Feature flags can act as permissions. They can allow or deny access to features to certain users. They act as a gate anywhere in the code, that you can turn on or off quickly, without changes to the code.

Reviewer: ```suggestion
Feature flags can act as permissions, granting or denying certain users access to features. They act as a gate anywhere in the code, that you can turn on or off quickly, without changes to the code.
```

> Instead of needing to have toggles inside your app to enable access to certain areas, you can set them up as feature flags. This enables teams to have quick and easy permissions. No need for custom user fields, modification in an admin panel, and additional checks, you can use feature flags and all the data you already have.

Reviewer: ```suggestion
Replacing in-app toggles with feature flags eliminates the need for custom user fields, admin panel modifications, and additional checks, granting teams quick, easy and flexible permissions.
```

Complex permission logic isn’t needed, engineering teams can just use a feature flag. This simplifies and distributes permission and access management. This means that a sales, support, or customer success team can then add permissions to their workflows and responsibilities. This takes work off the plate of developers. 

> Feature flags enable you to run alphas and betas of new features easily. Each feature doesn’t need a bespoke change to the code (other than the feature itself). The type of release or access pattern can change easily. For example, it can go from an alpha to a specific group release to a percentage rollout without needing to change any code.

Reviewer: ```suggestion
Feature flags also enable you to run alphas and betas of new features easily. Each feature doesn’t need a bespoke change to the code, other than the feature itself. The type of release or access pattern can change easily. For example, it can go from an alpha to a specific group release to a percentage rollout without needing to change any code.
```

> ### Use cases

Reviewer: ```suggestion
#### Use cases
```

> - Put premium features behind a permission-type feature flag. Once users sign up, get the sales or customer success team to add them to the feature flag distribution. If they churn, remove them from the feature flag release conditions.
> - A user has been abusing your app. They have been sending too many requests or using the feature inappropriately. You can use a permission-type feature flag to prevent or throttle access to key parts of your app from them.

Reviewer: ```suggestion
- Put premium features behind a permission-type feature flag. Once users sign up, get the sales or customer success team to add them to the feature flag distribution. If they churn, remove them from the feature flag release conditions.

- A user has been abusing your app. They have been sending too many requests or using the feature inappropriately. You can use a permission-type feature flag to prevent or throttle access to key parts of your app from them.
```

> 

Reviewer: I think this post would benefit from a summary section of feature flag tools. We don't need to go into huge depth, but it's a natural next step for the reader and an obvious opportunity to promote PostHog's features. 

Perhaps we could add a simplified version of the comparison table from https://posthog.com/product/feature-flags?

Author: Added, let me know what you think. I tried to tie categories to what was written about in the article combined with some of the key areas of differences between the tools.

Reviewer: Thoughts:

1. We should add a little commentary to this section – maybe a very short write-up on each the tools we mention, what they do, how they charge etc + and intro? It feels a bit abrupt without it. I should have said this, so my bad! We should add other features unique to us as well (e.g. session recording) to talk up our comprehensiveness. 

2. This is a fault of the original table imo, but maybe we should distinguish between "flag analytics" and "product analytics" in the table? Feels like we're giving Flagsmith and Growthbook a bit more credit than they deserve here!

3. After the table, do a short section on the benefit of having feature flags and product analytics all-in-one tool? This is a bit more openly marketing-y, but I think it's justified. Can end with some kind of CTA / link.

Author: Updated all three points. I think we are getting well beyond the scope of "Feature flag benefits, types and use cases, explained" if we add any more.

Reviewer: Agreed. This is just right.
